src/libcore/alloc.rs

```diff
@@ -120,14 +120,14 @@ impl Layout {
 
     /// Constructs a `Layout` suitable for holding a value of type `T`.
     #[stable(feature = "alloc_layout", since = "1.28.0")]
+    #[rustc_const_stable(feature = "alloc_layout_const_new", since = "1.42.0")]
     #[inline]
-    pub fn new<T>() -> Self {
+    pub const fn new<T>() -> Self {
         let (size, align) = size_align::<T>();
         // Note that the align is guaranteed by rustc to be a power of two and
         // the size+align combo is guaranteed to fit in our address space. As a
         // result use the unchecked constructor here to avoid inserting code
         // that panics if it isn't optimized well enough.
-        debug_assert!(Layout::from_size_align(size, align).is_ok());
```

Comment:
Now that we have conditionals in `const`, can we re-add the `debug_assert!`?

Reply:
Given that the stdlib is always compiled in release mode, it's not too useful to re-add.

But yes, given `feature(const_fn, const_if_match,` and `const_panic)`, this debug assert can be made to compile. (Ideally it'd also have `const fn Result::is_ok` to avoid using `matches!` directly.)

Reply:
We have some CI runners that have debug assertions enabled; it is not uncommon for rustc devs to enable debug assertions locally; Miri will hopefully enable libstd debug assertions eventually. So, there is some use to these (and the libstd is using them in plenty of places).

Reply:
`is_ok` was recently made const in #67685

Reply:
@RalfJung For some reason, while I [can get `debug_assert!` in `const_fn` on the playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a113bb0ff65efe9d01353d8b2160dc43), any attempt to add it in libcore, even `#[cfg(not(bootstrap))]` is giving me `error[E0723]: loops and conditional expressions are not stable in const fn`.

```diff
diff --git a/src/libcore/alloc.rs b/src/libcore/alloc.rs
index a04e75bc7ce..a679fe160a8 100644
--- a/src/libcore/alloc.rs
+++ b/src/libcore/alloc.rs
@@ -131,6 +131,8 @@ impl Layout {
         // the size+align combo is guaranteed to fit in our address space. As a
         // result use the unchecked constructor here to avoid inserting code
         // that panics if it isn't optimized well enough.
+        #[cfg(not(bootstrap))]
+        debug_assert!(Layout::from_size_align(size, align).is_ok());
         unsafe { Layout::from_size_align_unchecked(size, align) }
     }
 
diff --git a/src/libcore/lib.rs b/src/libcore/lib.rs
index bca96b77812..00ddbb923e2 100644
--- a/src/libcore/lib.rs
+++ b/src/libcore/lib.rs
@@ -82,6 +82,7 @@
 #![feature(const_int_pow)]
 #![feature(constctlz)]
 #![feature(const_panic)]
+#![feature(const_fn)]
 #![feature(const_fn_union)]
 #![feature(const_generics)]
 #![feature(const_ptr_offset_from)]
```

```
error[E0723]: loops and conditional expressions are not stable in const fn                                                                    
   --> src\libcore\macros\mod.rs:184:23
    |
183 | / macro_rules! debug_assert {
184 | |     ($($arg:tt)*) => (if $crate::cfg!(debug_assertions) { $crate::assert!($($arg)*); })
    | |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
185 | | }
    | |_- in this expansion of `debug_assert!`
    | 
   ::: src\libcore\alloc.rs:135:9
    |
135 |           debug_assert!(Layout::from_size_align(size, align).is_ok());
    |           ------------------------------------------------------------ in this macro invocation
    |
    = note: see issue #57563 <https://github.com/rust-lang/rust/issues/57563> for more information
    = help: add `#![feature(const_fn)]` to the crate attributes to enable
```

Reply:
@CAD97 Stable `const fn` cannot use unstable features. There's `allow_internal_unstable` to work around this, but at this point we might not yet want to stabilize methods that use these features internally.

Also @oli-obk is the better person for such questions. :) Or just ping the entire team: `@rust-lang/wg-const-eval`

Reply:
Actually looks like we [already have stable `const fn` that opt into using conditionals](https://doc.rust-lang.org/nightly/src/core/num/mod.rs.html#1810), so just adding `#[allow_internal_unstable(const_if_match)]` should be fine.

Reply:
It's not fine. The stable `const fn`s that use conditionals are just refactorings of hacks that were written without conditionals before, but they do not use more expressive power than what is available on stable, which, as far as I can tell, this would.

Reply:
This would just use it for a debug assertion though, which we can remove again any time if we have to.